File: scripts/flood_mapping.py
"""
Sentinel-1 SAR flood mapping — Python port of scripts/flood_mapping.js
using the earthengine-api. Produces the same flood extent, exports GeoTIFFs
locally via geemap, and writes the exposure table as CSV.

Setup:
    pip install earthengine-api geemap pandas
    earthengine authenticate   # one-time browser auth, same GEE account as the JS editor

Run:
    python flood_mapping.py
"""

import ee
import geemap
import pandas as pd

ee.Initialize()

# ---------- 1. STUDY AREA & DATES (EDIT THESE) ----------
GEOMETRY = ee.Geometry.Rectangle([76.0, 9.5, 77.5, 10.5])  # placeholder: Kerala backwaters
BEFORE_START, BEFORE_END = "2018-07-01", "2018-07-31"
AFTER_START, AFTER_END = "2018-08-15", "2018-08-25"
DIFF_THRESHOLD = 1.25
SMOOTHING_RADIUS = 50  # meters


def load_s1_collection():
    return (
        ee.ImageCollection("COPERNICUS/S1_GRD")
        .filter(ee.Filter.eq("instrumentMode", "IW"))
        .filter(ee.Filter.listContains("transmitterReceiverPolarisation", "VH"))
        .filter(ee.Filter.eq("orbitProperties_pass", "DESCENDING"))
        .filterBounds(GEOMETRY)
        .select("VH")
    )


def compute_flood_extent():
    collection = load_s1_collection()
    before = collection.filterDate(BEFORE_START, BEFORE_END).mosaic().clip(GEOMETRY)
    after = collection.filterDate(AFTER_START, AFTER_END).mosaic().clip(GEOMETRY)

    before_f = before.focal_median(SMOOTHING_RADIUS, "circle", "meters")
    after_f = after.focal_median(SMOOTHING_RADIUS, "circle", "meters")

    difference = after_f.divide(before_f)
    flooded = difference.gt(DIFF_THRESHOLD).rename("water").selfMask()

    permanent_water = (
        ee.Image("JRC/GSW1_4/GlobalSurfaceWater").select("seasonality").gte(10).clip(GEOMETRY)
    )
    flooded = flooded.updateMask(permanent_water.Not())

    dem = ee.Image("WWF/HydroSHEDS/03VFDEM")
    slope = ee.Terrain.slope(dem)
    flooded = flooded.updateMask(slope.lt(5))

    connections = flooded.connectedPixelCount()
    flooded_clean = flooded.updateMask(connections.gt(8))

    return flooded_clean, before_f, after_f, difference


def area_hectares(mask_image, geometry=GEOMETRY, scale=10):
    area_img = mask_image.select("water").multiply(ee.Image.pixelArea())
    stats = area_img.reduceRegion(
        reducer=ee.Reducer.sum(), geometry=geometry, scale=scale, maxPixels=1e13
    )
    m2 = stats.get("water").getInfo() or 0
    return m2 / 10000.0


def exposure_table(flooded_clean):
    worldcover = ee.ImageCollection("ESA/WorldCover/v200").first().clip(GEOMETRY)
    cropland = worldcover.eq(40)
    builtup = worldcover.eq(50)

    cropland_ha = area_hectares(flooded_clean.updateMask(cropland))
    builtup_ha = area_hectares(flooded_clean.updateMask(builtup))

    return pd.DataFrame(
        [
            {"category": "cropland", "flooded_hectares": cropland_ha},
            {"category": "built-up", "flooded_hectares": builtup_ha},
        ]
    )


def export_geotiff(image, filename, geometry=GEOMETRY, scale=10):
    geemap.ee_export_image(
        image, filename=filename, scale=scale, region=geometry, file_per_band=False
    )


if __name__ == "__main__":
    flooded_clean, before_f, after_f, difference = compute_flood_extent()

    total_ha = area_hectares(flooded_clean)
    print(f"Total flooded area: {total_ha:,.1f} hectares")

    table = exposure_table(flooded_clean)
    table.to_csv("../exports/exposure_table.csv", index=False)
    print(table)

    export_geotiff(flooded_clean, "../exports/flood_extent.tif")
    export_geotiff(before_f, "../exports/before_vh.tif")
    export_geotiff(after_f, "../exports/after_vh.tif")

File: scripts/flood_mapping.js
/*
  Sentinel-1 SAR flood mapping, Kerala floods (August 2018).
  Change-detection method based on the UN-SPIDER recommended practice:
  https://un-spider.org/advisory-support/recommended-practices/recommended-practice-google-earth-engine-flood-mapping
*/

// ---------- 1. STUDY AREA & DATES ----------
// Kuttanad, Alappuzha district — the low-lying "rice bowl" of Kerala (parts sit
// below sea level), among the worst-hit areas in the August 2018 floods.
var geometry = ee.Geometry.Rectangle([76.25, 9.25, 76.65, 9.75]);

var before_start = '2018-07-01';
var before_end   = '2018-07-31';   // dry-season reference window, same year
var after_start  = '2018-08-15';
var after_end    = '2018-08-25';   // flood event window

Map.centerObject(geometry, 8);

// ---------- 2. LOAD SENTINEL-1 GRD, FILTER ----------
var collection = ee.ImageCollection('COPERNICUS/S1_GRD')
  .filter(ee.Filter.eq('instrumentMode', 'IW'))
  .filter(ee.Filter.listContains('transmitterReceiverPolarisation', 'VH'))
  .filter(ee.Filter.eq('orbitProperties_pass', 'DESCENDING')) // keep before/after on same pass
  .filterBounds(geometry)
  .select('VH');

var beforeCollection = collection.filterDate(before_start, before_end);
var afterCollection  = collection.filterDate(after_start, after_end);

var before = beforeCollection.mosaic().clip(geometry);
var after  = afterCollection.mosaic().clip(geometry);

// ---------- 3. SPECKLE FILTERING ----------
// Focal median smoothing — a simpler stand-in for a proper Refined Lee filter.
var smoothing_radius = 50; // meters
var before_filtered = before.focal_median(smoothing_radius, 'circle', 'meters');
var after_filtered  = after.focal_median(smoothing_radius, 'circle', 'meters');

// ---------- 4. CHANGE DETECTION ----------
var difference = after_filtered.divide(before_filtered);

// Fixed threshold (UN-SPIDER default) — otsu_threshold.js computes an alternative.
var diff_threshold = 1.25;
var flooded = difference.gt(diff_threshold).rename('water').selfMask();

// ---------- 5. MASK OUT PERMANENT WATER (JRC Global Surface Water) ----------
var permanentWater = ee.Image('JRC/GSW1_4/GlobalSurfaceWater')
  .select('seasonality').gte(10).clip(geometry);
var flooded_masked = flooded.updateMask(permanentWater.not());

// ---------- 6. MASK STEEP SLOPES (removes radar shadow/layover false positives) ----------
var dem = ee.Image('WWF/HydroSHEDS/03VFDEM');
var slope = ee.Terrain.slope(dem);
var flooded_final = flooded_masked.updateMask(slope.lt(5));

// ---------- 7. REMOVE ISOLATED PIXELS (connectivity filter) ----------
var connections = flooded_final.connectedPixelCount();
var flooded_clean = flooded_final.updateMask(connections.gt(8));

// ---------- 8. AREA CALCULATION ----------
var flood_pixelarea = flooded_clean.select('water').multiply(ee.Image.pixelArea());
var flood_stats = flood_pixelarea.reduceRegion({
  reducer: ee.Reducer.sum(),
  geometry: geometry,
  scale: 10,
  maxPixels: 1e13
});
print('Flooded area (m^2):', flood_stats.get('water'));
print('Flooded area (hectares):', ee.Number(flood_stats.get('water')).divide(10000));

// ---------- 9. VISUALIZATION ----------
Map.addLayer(before_filtered, {min: -25, max: 0}, 'Before (VH dB)', false);
Map.addLayer(after_filtered, {min: -25, max: 0}, 'After (VH dB)', false);
Map.addLayer(flooded_clean, {palette: ['#00FFFF']}, 'Flood extent');

// ---------- 10. EXPORT ----------
Export.image.toDrive({
  image: flooded_clean,
  description: 'flood_extent',
  folder: 'sar-flood-mapping',
  region: geometry,
  scale: 10,
  maxPixels: 1e13
});

// Vectorize the flood mask so it can be dropped into the Streamlit app as a
// lightweight GeoJSON (no raster tile server needed to display it).
var flood_vectors = flooded_clean.reduceToVectors({
  geometry: geometry,
  scale: 10,
  geometryType: 'polygon',
  eightConnected: false,
  maxPixels: 1e13
});
Export.table.toDrive({
  collection: flood_vectors,
  description: 'flood_extent_vector',
  folder: 'sar-flood-mapping',
  fileFormat: 'GeoJSON'
});
